### ISSUE TITLE
Current testname tracking

### ISSUE THREAD

**Issue by mpettigr** [svunit/svunit#54]
svunit is great and it's a remarkable boost in my rtl design development. Thank you!

In using it for the last few months, I would like to mention a frequent pain point and my workaround.

Unit test files can each contain numerous tests.  It would be great to have a way to only call a single test or group from the file via a regex or similar at the CLI.  In the meantime, I have to run all the tests and often review the results in my waveform viewer.  

Figuring out where in the sim a test ends and the next begins is a challenge so an svunit idiom I now always use in all of my svunit test files as a workaround is the following:

1. Add string current_testname = "" after the svunit_testcase declaration, i.e.,

`module foobar_unit_test;

   import svunit_pkg::svunit_testcase;

   
   string name = "foobar_ut";
   svunit_testcase svunit_ut;

   string current_testname = "";
`
2. Inside each `SVTEST() block add the following:
`     `SVTEST( test_it )

// add signal declarations here
   bit [31:0] test_foo;
   int test_expected_data10][2];
   
   current_testname = _testName;`

This has the effect of updating current_testname variable at the start of each test.  If you wave current_testname in your waveform viewer, it's really easiest to see which test you are in and you can use your waveform viewer's search function to hop to the desired test.  This is really helpful and wanted to share it with you and other svunit users.

I don't think there is a way to automate this but it would be great if there were.

Thanks again Neil for making such a great unit test framework.

-Mark

**Comment by mpettigr** [comment]
Actually, if you desired to automate the inclusion of a current_testname you could do the following:

1. Update svunit test generator to include the module variable declaration of:
`string current_testname = "";`
2. Update SVUNIT(_NAME_) macro
and just add
`current_testname = _testName` 

to it.

You could potentially just reuse _testName but it's not visible at the module level where you have it for easy add to waveform viewer because it's a task variable (with a narrower scope).

**Comment by nosnhojn** [comment]
just to make sure I have this right, it sounds like you've got a couple suggestions here...
(a) add a string property so you can see in a waveform which test you're in; and
(b) add a command line option so that you could run exactly 1 test and avoid running the rest

did I get that right?

I think (a) is pretty simple so no problem there. it's a good suggestion. re: (b) would it be adequate to run with the -t <*unit_test.sv> to cut your sim down to only 1 unit_test file then use the test string to find your test? or would it be outstanding to be just running 1 test and ignoring the rest?

**Comment by tudortimi** [comment]
I think we could probably also build option b). I've found myself wanting something similar a lot of times. We would need an extra command line option that takes the name of the SVUNIT_TEST to run and just compare against this name in the runner, after the simulation has started. It would only be allowed to specify a single test name like this if it is unique across all tests in the run, otherwise we could issue an error and ask for an extra '-t' to narrow it down to a single 'unit_test.sv' file.

**Comment by nosnhojn** [comment]
@tudortimi you think it makes more sense to overload the '-t <blah>' you added or add a new switch? Because the (b) functionality can't coexist with '-t', part of me wants to overload '-t' to minimize the number of switches. but only if it's obvious to users.

**Comment by tudortimi** [comment]
@nosnhojn I don't think that this functionality can't coexist with '-t'; quite the opposite. '-t' is useful to run all testcases for a unit. The extra switch would be useful to focus on a single testcase.

Imagine you want to work on a certain task. You write a new test. You want to focus on getting this test to pass. For this, the extra switch would be ideal. You write your production code that makes this test pass. You now want to make sure that you didn't break anything in your unit. You run with '-t' and specify only the unit test file for this unit. Lastly, you might also want to run all unit tests (this isn't as important if you follow the 'mockist' style).

**Comment by nosnhojn** [comment]
@tudortimi so I was thinking the '-t <arg>' would look first for a unit test template. if <arg> is a unit_test.sv, then it runs every test in that _unit_test.sv. if not, it'll run the SVTEST(<arg>). I guess a fail-safe is that if no tests are run, we exit with an error assuming it wasn't intentional.

**Comment by tudortimi** [comment]
SVTEST names aren't unique across all unit test files. If you want to reuse the '-t' switch, then we could use a syntax like '-t my_unit_test.some_func__some_condition__does_stuff', where 'my_unit_test' is the unit test file and 'some_func__some_condition__does_stuff' is the name of an SVTEST defined within that file.

**Comment by tudortimi** [comment]
This is similar to how Python does it:

```
python -m unittest test_module1 test_module2
python -m unittest test_module.TestClass
python -m unittest test_module.TestClass.test_method
```

Source: [https://docs.python.org/2/library/unittest.html](https://docs.python.org/2/library/unittest.html)

**Comment by nosnhojn** [comment]
yes. I like that.

@mpettigr does this make sense to you? decent solution for what you're looking for?
